stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -561,13 +561,24 @@ public struct Set<Element : Hashable> :
   public func intersect<
     S : SequenceType where S.Generator.Element == Element
   >(sequence: S) -> Set<Element> {
-    let other = sequence as? Set<Element> ?? Set(sequence)
-    var newSet = Set<Element>()
-    for member in self {
-      if other.contains(member) {
-        newSet.insert(member)
+
+    // Attempt to iterate the smaller between `self` and `sequence`.
+    // If sequence is not a set, iterate over it because it may be single-pass.
+    var newSet: Set<Element>
+
+    if (sequence as? Set<Element>)?.count > count {
+      newSet = Set<Element>(minimumCapacity: count)
+      for element in self where sequence.contains(element) {
+        newSet.insert(element)
+      }
+    } else {
+      // sequence is ether a Set with a smaller count or not a Set at all.
+      newSet = Set<Element>(minimumCapacity: sequence.underestimateCount())
```

Comment:
I think it's better to use `min(sequence.underestimateCount(), count)` for `minimumCapacity`. In that way you don't allocate too much memory for the new set.